crates/sema/src/stats.rs

```diff
@@ -0,0 +1,243 @@
+        node.stats.count += 1;
+        node.stats.size = std::mem::size_of_val(val);
+
+        if let Some(label2) = label2 {
+            let subnode = node.subnodes.entry(label2).or_insert(NodeStats::new());
+            subnode.count += 1;
+            subnode.size = std::mem::size_of_val(val);
+        }
+    }
+
+    fn print(&self, title: &str, prefix: &str) {
+        let mut nodes: Vec<_> = self.nodes.iter().collect();
+        nodes.sort_by_key(|(_, node)| node.stats.count * node.stats.size);
+
+        let total_size = nodes.iter().map(|(_, node)| node.stats.count * node.stats.size).sum();
+
+        eprintln!("{prefix} {title}");
+        eprintln!(
+            "{} {:<18}{:>18}{:>14}{:>14}",
+            prefix, "Name", "Accumulated Size", "Count", "Item Size"
+        );
+        eprintln!("{prefix} ----------------------------------------------------------------");
+
+        let percent = |m, n| (m * 100) as f64 / n as f64;
+
+        for (label, node) in nodes {
+            let size = node.stats.count * node.stats.size;
+            eprintln!(
+                "{} {:<18}{:>10} ({:4.1}%){:>14}{:>14}",
+                prefix,
+                label,
+                to_readable_str(size),
+                percent(size, total_size),
+                to_readable_str(node.stats.count),
+                to_readable_str(node.stats.size)
+            );
+            if !node.subnodes.is_empty() {
+                // We will soon sort, so the initial order does not matter.
+                #[allow(rustc::potential_query_instability)]
+                let mut subnodes: Vec<_> = node.subnodes.iter().collect();
+                subnodes.sort_by_key(|(_, subnode)| subnode.count * subnode.size);
+
+                for (label, subnode) in subnodes {
+                    let size = subnode.count * subnode.size;
+                    eprintln!(
+                        "{} - {:<18}{:>10} ({:4.1}%){:>14}",
+                        prefix,
+                        label,
+                        to_readable_str(size),
+                        percent(size, total_size),
+                        to_readable_str(subnode.count),
+                    );
+                }
+            }
+        }
+        eprintln!("{prefix} ----------------------------------------------------------------");
+        eprintln!("{} {:<18}{:>10}", prefix, "Total", to_readable_str(total_size));
+        eprintln!("{prefix}");
+    }
+}
+
+// Used to avoid boilerplate for types with many variants.
+macro_rules! record_variants {
+    (
+        ($self:ident, $val:expr, $kind:expr, $id:expr, $mod:ident, $ty:ty, $tykind:ident),
+        [$($variant:ident),*]
+    ) => {
+        match $kind {
+            $(
+                $mod::$tykind::$variant { .. } => {
+                    $self.record_variant(stringify!($ty), stringify!($variant), $id, $val)
+                }
+            )*
+        }
+    };
+}
+
+impl<'ast> Visit<'ast> for StatCollector {
```

Comment:
I'm probably missing some cases here...

If I understand correctly, I think I need to recurse in the cases where an AST element is being stored within a parent node, and not in the cases where it's just a reference (i.e. not taking up additional memory)

Reply:
Yes I believe that's right

Reply:
I did some hand tuning -- basically when an AST struct contains another visitable thing if it's a Boxed reference to that thing, then we visit it (so we can uncover the size of that heap-allocated structure), but if it's inlined (e.g. Ident, Span, etc.), I have the code skip the field since it that Ident or Span or other data structure is already going to be accounted for when `std::mem::size_of_val` is called on the parent value.